### PR TITLE
Merge dev branches

### PR DESCRIPTION
**Pull request by LBern**
## General

Updating `devel/alice` and `develop` to be on the same page again.

### CODE REVIEW

File: visualization/include/visualization/point-cloud-filter.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  pcl::VoxelGrid<T_point> voxel_grid_;
+  voxel_grid_.setLeafSize(vg_leaf_size_m, vg_leaf_size_m, vg_leaf_size_m);
+  voxel_grid_.setInputCloud(cloud);
+  voxel_grid_.filter(*cloud);
+}
+
+template <typename T_point>
+void beautifyPointCloud(
+    PclPointCloudPtr<T_point> cloud, sensor_msgs::PointCloud2* cloud_msg) {
+  CHECK_NOTNULL(cloud_msg);
```
Comment (smauq) [inline visualization/include/visualization/point-cloud-filter.h:33]: You should also `CHECK_NOTNULL(cloud)`

File: visualization/src/resource-visualization.cc [hunk 2]
```diff
@@ -63,8 +64,51 @@ DEFINE_bool(
...
+  if (pcl_cloud->empty()) {
+    return;
+  }
+  voxelGridPointCloud(FLAGS_vis_pointcloud_filter_leaf_size_m, pcl_cloud);
+  backend::convertPointCloudType((*pcl_cloud), cloud_filtered);
+}
+
+template <typename T_input>
+static void applyBeautification(
+    const T_input& cloud_in, sensor_msgs::PointCloud2* cloud_filtered) {
```
Comment (smauq) [inline visualization/src/resource-visualization.cc:99]: Also check `CHECK_NOTNULL(cloud_filtered)`?

File: visualization/src/resource-visualization.cc [hunk 3]
```diff
@@ -256,31 +306,55 @@ void visualizeReprojectedDepthResource(
...
+        if (FLAGS_vis_pointcloud_filter_dense_map_before_publishing) {
+          pcl::PointCloud<pcl::PointXYZI>::Ptr pcl_cloud(
+              new pcl::PointCloud<pcl::PointXYZI>);
+          CHECK_NOTNULL(pcl_cloud);
+          backend::convertPointCloudType(points_S, &(*pcl_cloud));
+          /*
+          voxelGridPointCloud(
+              FLAGS_vis_pointcloud_filter_leaf_size_m, pcl_cloud);
+          backend::convertPointCloudType((*pcl_cloud), &points_S);
+          */
```
Comment (smauq) [inline visualization/src/resource-visualization.cc:318]: Keep or remove this commented out bit?

File: visualization/src/resource-visualization.cc [hunk 4]
```diff
@@ -256,31 +306,55 @@ void visualizeReprojectedDepthResource(
           return;
         }
 
+        if (FLAGS_vis_pointcloud_filter_dense_map_before_publishing) {
+          pcl::PointCloud<pcl::PointXYZI>::Ptr pcl_cloud(
```
Comment (smauq) [inline visualization/src/resource-visualization.cc:310]: This memory isn't deallocated and leaks

File: visualization/src/resource-visualization.cc [hunk 5]
```diff
@@ -256,31 +306,55 @@ void visualizeReprojectedDepthResource(
           return;
         }
 
+        if (FLAGS_vis_pointcloud_filter_dense_map_before_publishing) {
```
Comment (smauq) [inline visualization/src/resource-visualization.cc:309]: Overall I can't figure that this if statement does anything, should this be removed completely?

File: visualization/src/resource-visualization.cc [hunk 6]
```diff
@@ -256,31 +306,55 @@ void visualizeReprojectedDepthResource(
...
             sensor_msgs::PointCloud2 ros_point_cloud_S;
             backend::convertPointCloudType(
                 points_C_colorized, &ros_point_cloud_S);
-            publishPointCloudInLocalFrameWithTf(T_G_S, &ros_point_cloud_S);
           } else {
-            sensor_msgs::PointCloud2 ros_point_cloud_S;
             backend::convertPointCloudType(points_S, &ros_point_cloud_S);
-            publishPointCloudInLocalFrameWithTf(T_G_S, &ros_point_cloud_S);
           }
+          if (FLAGS_vis_pointcloud_filter_dense_map_before_publishing) {
```
Comment (smauq) [inline visualization/src/resource-visualization.cc:333]: Here I am confused by the flow of the if statements. From the config it's not obvious to me that `FLAGS_vis_pointcloud_filter_dense_map_before_publishing` and `FLAGS_vis_pointcloud_filter_beautify_dense_map_before_publishing` are exclusive, but they seem to be so. If they are exclusive, there should be a check and warning for this. If they are meant to work together this needs to be reworked.

File: visualization/src/resource-visualization.cc [hunk 7]
```diff
@@ -256,31 +306,55 @@ void visualizeReprojectedDepthResource(
...
+            applyVoxelGridFilter(ros_point_cloud_S, &ros_point_cloud_S);
+          } else if (
+              FLAGS_vis_pointcloud_filter_beautify_dense_map_before_publishing) {  // NOLINT
+            applyBeautification(points_S, &ros_point_cloud_S);
```
Comment (smauq) [inline visualization/src/resource-visualization.cc:337]: This drops the coloring previously done. Intentional?

Comment (smauq) [inline visualization/src/resource-visualization.cc:337]: Additionally the flow of the entire `applyBeautification` function is unclear to me, it seems to drop the content of `ros_point_cloud_S` totally and just replaces it based on `points_S`. Then all the previous work is useless always

File: visualization/src/resource-visualization.cc [hunk 8]
```diff
@@ -256,31 +306,55 @@ void visualizeReprojectedDepthResource(
...
           }
 
           sensor_msgs::PointCloud2 ros_point_cloud_G;
-          backend::convertPointCloudType(points_G, &ros_point_cloud_G);
-
+          if (FLAGS_vis_pointcloud_filter_dense_map_before_publishing) {
+            applyVoxelGridFilter(points_G, &ros_point_cloud_G);
+          } else if (
+              FLAGS_vis_pointcloud_filter_beautify_dense_map_before_publishing) {  // NOLINT
+            applyBeautification(points_G, &ros_point_cloud_G);
```
Comment (smauq) [inline visualization/src/resource-visualization.cc:354]: See my comment above about the flags being exclusive.

File: visualization/src/resource-visualization.cc [hunk 9]
```diff
@@ -304,7 +378,14 @@ void visualizeReprojectedDepthResource(
 
   // Publish accumulated point cloud in global frame.
   sensor_msgs::PointCloud2 ros_point_cloud_G;
-  backend::convertPointCloudType(accumulated_point_cloud_G, &ros_point_cloud_G);
+  if (FLAGS_vis_pointcloud_filter_dense_map_before_publishing) {
+    applyVoxelGridFilter(accumulated_point_cloud_G, &ros_point_cloud_G);
+  } else if (FLAGS_vis_pointcloud_filter_beautify_dense_map_before_publishing) {
+    applyBeautification(accumulated_point_cloud_G, &ros_point_cloud_G);
```
Comment (smauq) [inline visualization/src/resource-visualization.cc:384]: See my comment above about the flags being exclusive.

File: visualization/src/resource-visualization.cc [hunk 10]
```diff
@@ -410,10 +491,24 @@ void visualizeReprojectedDepthResourcePerRobot(
...
+
     // Publish accumulated point cloud in global frame.
     sensor_msgs::PointCloud2 ros_point_cloud_G;
-    backend::convertPointCloudType(
-        accumulated_point_cloud_G, &ros_point_cloud_G);
+    if (FLAGS_vis_pointcloud_filter_dense_map_before_publishing) {
+      applyVoxelGridFilter(accumulated_point_cloud_G, &ros_point_cloud_G);
+    } else if (
+        FLAGS_vis_pointcloud_filter_beautify_dense_map_before_publishing) {
+      applyBeautification(accumulated_point_cloud_G, &ros_point_cloud_G);
```
Comment (smauq) [inline visualization/src/resource-visualization.cc:506]: See my comment above about the flags being exclusive.

**Review by smauq — CHANGES_REQUESTED** [review]
First pass over the things. I left a bunch of comments. The new lidar stuff looks cool!

File: .gitmodules [hunk 11]
```diff
@@ -100,6 +100,9 @@
 [submodule "dependencies/3rdparty/pointmatcher-ros"]
 	path = dependencies/3rdparty/pointmatcher-ros
 	url = git@github.com:ANYbotics/pointmatcher-ros.git
-[submodule "dependencies/internal/libpointmatcher"]
-	path = dependencies/internal/libpointmatcher
-	url = git@github.com:ethz-asl/libpointmatcher.git
+[submodule "dependencies/3rdparty/libpointmatcher"]
```
[the remark this answers is not in the document]
Comment (LBern) [reply .gitmodules:103]: we had problems on 20.04 with the other version

File: algorithms/landmark-triangulation/include/landmark-triangulation/pose-interpolator.h [hunk 12]
```diff
@@ -43,7 +43,7 @@ class PoseInterpolator {
   // imu_timestamps.  Timestamps do not need to be sorted, but must lie within
   // the minimum and maximum timestamp of the mission's IMU measurements.
   // Extrapolating outside this range is not supported.
-  void getPosesAtTime(
+  bool getPosesAtTime(
```
[the remark this answers is not in the document]
Comment (LBern) [reply algorithms/landmark-triangulation/include/landmark-triangulation/pose-interpolator.h:46]: Added checks to the callers.

File: algorithms/registration-toolbox/include/registration-toolbox/alignment/pcl-alignment.h [hunk 13]
```diff
@@ -45,32 +58,149 @@ RegistrationResult PclAlignment<T_alignment, T_point>::registerCloudImpl(
...
+  pcl::getMinMax3D(*cloud, min_point, max_point);
+  const std::int64_t dx = static_cast<std::int64_t>(
+                              (max_point.x - min_point.x) * inverse_grid_size) +
+                          1;
+  const std::int64_t dy = static_cast<std::int64_t>(
+                              (max_point.y - min_point.y) * inverse_grid_size) +
+                          1;
+  const std::int64_t dz = static_cast<std::int64_t>(
+                              (max_point.z - min_point.z) * inverse_grid_size) +
+                          1;
```
[the remark this answers is not in the document]
Comment (LBern) [reply algorithms/registration-toolbox/include/registration-toolbox/alignment/pcl-alignment.h:139]: @patripfr

File: algorithms/registration-toolbox/include/registration-toolbox/alignment/pcl-alignment.h [hunk 14]
```diff
@@ -45,32 +58,149 @@ RegistrationResult PclAlignment<T_alignment, T_point>::registerCloudImpl(
...
+  const std::int64_t dxdydz = dx * dy * dz;
+  // Find amount of levels that we have to split the pointcloud into
+  bool indices_fit_in_voxel_grid = false;
+  while (!indices_fit_in_voxel_grid) {
+    if ((dxdydz / (volume_level * volume_level * volume_level)) <
```
[the remark this answers is not in the document]
Comment (LBern) [reply algorithms/registration-toolbox/include/registration-toolbox/alignment/pcl-alignment.h:144]: @patripfr

File: common/resources-common/include/resources-common/point-cloud.h [hunk 15]
```diff
@@ -306,50 +319,57 @@ struct PointCloud {
...
+            // count.
+            CHECK_EQ(xyz_point_count, scalar_count);
+          }
 
-        if (scalar_count > 0) {
-          // If a value attribute is present, its count should match the point
-          // count.
-          CHECK_EQ(xyz_point_count, scalar_count);
-        }
+          if (label_count > 0) {
```
[the remark this answers is not in the document]
Comment (LBern) [reply common/resources-common/include/resources-common/point-cloud.h:356]: @patripfr

File: common/resources-common/src/point-cloud.cc [hunk 16]
```diff
@@ -0,0 +1,239 @@
...
+  std::ifstream input_file(file_path, std::ios::binary);
+  if (!input_file) {
+    LOG(ERROR) << "Failed opening the input file.";
+    return false;
+  }
+
+  // Read the file stream into a buffer.
+  std::streampos file_size = 0;
+  input_file.seekg(0, std::ios::end);
+  file_size = input_file.tellg() - file_size;
```
[the remark this answers is not in the document]
Comment (LBern) [reply common/resources-common/src/point-cloud.cc:111]: @patripfr

File: visualization/src/resource-visualization.cc [hunk 17]
```diff
@@ -63,8 +64,51 @@ DEFINE_bool(
...
+}
+
+template <typename T_input>
+static void applyBeautification(
+    const T_input& cloud_in, sensor_msgs::PointCloud2* cloud_filtered) {
+  pcl::PointCloud<pcl::PointXYZI>::Ptr pcl_cloud(
+      new pcl::PointCloud<pcl::PointXYZI>);
+  CHECK_NOTNULL(pcl_cloud);
+
+  backend::convertPointCloudType(cloud_in, &(*pcl_cloud));
```
[the remark this answers is not in the document]
Comment (LBern) [reply visualization/src/resource-visualization.cc:104]: If I remember correctly the issue was that `convertPointCloudType` wants a raw pointer of the type. that's why this workaround here.

File: visualization/src/resource-visualization.cc (hunk 4 above)
Comment (LBern) [reply visualization/src/resource-visualization.cc:310]: pcl pointers should be smart pointers

File: common/resources-common/include/resources-common/point-cloud.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (patripfr) [reply common/resources-common/include/resources-common/point-cloud.h:356]: @smauq what exactly do you mean here? we are checking for scalars as well in line 350

File: visualization/src/resource-visualization.cc (hunk 6 above)
Comment (LBern) [reply visualization/src/resource-visualization.cc:346]: `FLAGS_vis_pointcloud_filter_dense_map_before_publishing` is deprecated as we anyway can do this with the beautification. I removed the respective parts.

File: visualization/src/resource-visualization.cc (hunk 7 above)
Comment (LBern) [reply visualization/src/resource-visualization.cc:350]: fixed

File: common/resources-common/include/resources-common/point-cloud.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (smauq) [reply common/resources-common/include/resources-common/point-cloud.h:356]: @patripfr True, sorry I must have missed that

File: algorithms/registration-toolbox/include/registration-toolbox/alignment/pcl-alignment.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (patripfr) [reply algorithms/registration-toolbox/include/registration-toolbox/alignment/pcl-alignment.h:139]: for the unlikely case that `max_point.z - min_point.z == 0`  `std::ceil`  would result in `dz = 0`. But I think `std::max(std::ceil((max_point.z - min_point.z) *
                                  inverse_grid_size), 1)` does the trick and is more elegant

**Pull request closed without merging**